File: 003_longest_substring_with_no_repetition/Solution.java
import java.util.HashSet;

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int j = 0, max = 0;
        for (int i = 0; i < s.length(); i ++){
            char c = s.charAt(i);
            if (!set.contains(c)){
                set.add(c);
            }else{
                max = Math.max(max, i - j);
                while (s.charAt(j) != c){
                    set.remove(s.charAt(j));
                    j++;
                }
                j++;
            }
        }
        return Math.max(max, s.length() - j);
    }


	public static void main(String[] args) {
		IO in = new IO();
		Solution sol = new Solution();
		while (in.hasNext()) {
			String s = in.nextLine();
			int ans = sol.lengthOfLongestSubstring(s);
			System.out.println(ans);
		}
	}
}

File: 003_longest_substring_with_no_repetition/Solution.cpp
#include <unordered_set>
#include <string>

#include "../IO.h"

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> set;
        int j = 0, max = 0;
        for (int i = 0; i < s.length(); i++){
            char c = s.at(i);
            if (set.find(c) == set.end()){
                set.insert(c);
            } else {
                if (i - j > max)
                    max = i - j;
                while (s.at(j) != c){
                    set.erase(s.at(j));
                    j++;
                }
                j++;
            }
        }
        if (s.length() - j > max)
            max = s.length() - j;
        return max;
    }
};

int main (int argc, char** argv) {
  Solution sol;
  while(true){
  	string s;
    cin >> s;
    if (!cin)
      break;
    int ans = sol.lengthOfLongestSubstring(s);
    cout << ans << endl;
  }
  return 0;
}



